Eran/binomial functions

eranamar opened the pull request:
Adding the binomial distribution, which is central to many statistical procedures (https://en.wikipedia.org/wiki/Binomial_distribution) (#15812 )

Test plan (adding unit-tests)

Following the diff template of: https://github.com/prestodb/presto/pull/11981/files

```
== RELEASE NOTES ==

General Changes
* Add :func:`binomial_cdf` and :func:`inverse_binomial_cdf` functions.
```

(like the beta_cds: https://prestodb.io/docs/current/release/release-0.215.html?highlight=beta_cds)

leepface commented:
Code review:
- [ ] The maven-checks errors are code formatting. If you can format in intellij (no need to have a working build environment), those should go away.
- [ ] Looks like tests may separately be failing
- [ ] Squash commits together

eranamar commented:
@leepface I've reformat the code. not sure how to squash all commits on my side. I thought there is a checkbox "squash commits" within the pull request but I can't find it.

leepface commented:
@eranamar You should be able to squash using this: https://stackoverflow.com/a/5201642. I think only the repo owner can squash from the PR.

eranamar commented:
@leepface I addressed all the code-review comments. Currently the pipeline status is "All checks have passed". 
Thanks for your valuable comments, I appreciate it. Please let me know if there is anything else.

talgalili commented:
Thanks @eranamar for all the changes, I think your diff is now ready for review (/merger).

rongrong commented:
Please squash all changes into a single commit and use `git push -f` to update the PR. Thanks!

eranamar commented:
@rongrong done :) 
I appreciate the time you put into reviewing my PR. Thank you very much!

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:
```diff
@@ -216,6 +217,20 @@ public static double atan2(@SqlType(StandardTypes.DOUBLE) double num1, @SqlType(
         return Math.atan2(num1, num2);
     }
 
+    @Description("Binomial cdf given n, prob, and a value")
```
what is n? I assume `prob` is `probability`? Can we make it more novice reader friendly (also the variable names).

eranamar commented:
@rongrong I've renamed the variables.

talgalili commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:
```diff
@@ -503,6 +518,21 @@ public static long floorFloat(@SqlType(StandardTypes.REAL) long num)
 
+    @Description("inverse of Binomial cdf given numberOfTrials, successProbability parameters and p")
+    @ScalarFunction
+    @SqlType(StandardTypes.INTEGER)
```
why not BIGINT?

eranamar replied:
number of trials is an integer, so how the result can be larger than an integer? (if you meant why at all integer rather than big-int, I guess that I just mimicked the types of the underline java implementation)

talgalili replied:
Interesting. I wonder if it can handle BIGINT or only int. Do you happen to know how to check that?

talgalili replied:
Update: after checking, it seems that Java's int is 4 bytes (like Presto's INTEGER) - so what you did is good.

talgalili commented on presto-docs/src/main/sphinx/functions/math.rst:
```diff
@@ -66,6 +72,13 @@ Mathematical Functions
 
     Returns the value of ``string`` interpreted as a base-``radix`` number.
 
+.. function:: inverse_binomial_cdf(numberOfTrials, successProbability, p) -> int
```
if to change to bigint, probably also to mention it here.

eranamar commented:
@rongrong Is there a way to rerun the tests? I have run `./mvnw test -fn -pl presto-cache` locally with my changes and the tests have passed. Alternatively, how can I debug this \ see the error trace?

talgalili commented:
You can rebase and push again. That should rerun the tests.

On Thu, Apr 29, 2021, 11:28 Eran Amar ***@***.***> wrote:

> @rongrong <https://github.com/rongrong> Is there a way to rerun the
> tests? I have run ./mvnw test -fn -pl presto-cache locally with my
> changes and the tests have passed. Alternatively, how can I debug this \
> see the error trace?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/15812#issuecomment-829041555>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHOJBXFNBX3AU2JPJWWEETTLEKB5ANCNFSM4ZACV2GA>
> .
>

eranamar commented:
@rongrong is there anything else I need to do with my PR? 
please let me know.

rongrong commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:
```diff
@@ -503,6 +518,21 @@ public static long floorFloat(@SqlType(StandardTypes.REAL) long num)
+            @SqlType(StandardTypes.DOUBLE) double successProbability,
+            @SqlType(StandardTypes.DOUBLE) double p)
+    {
+        checkCondition(p >= 0 && p <= 1, INVALID_FUNCTION_ARGUMENT, "p must be in the interval [0, 1]");
```
Nits: you are using 0.0 and 1.0 in `binomialCdf` but integer here. Please make them consistent (and consistent with other functions in the class). Thanks!

eranamar replied:
Done

rongrong reviewed (commented):
Looks good. Just one nit.

rongrong commented:
Also please change the commit title to "Add binomial distribution functions" (remove the unit tests part).

eranamar commented:
@rongrong I've done the following:
* make the change you suggested and become consistent with other classes (e.g. Beta distribution)
* rebased on latest and renamed the commit
Please let me know if I missed something.
Thank you very much for the valuable comments :)
